domain/port/state/state.go

```diff
@@ -0,0 +1,271 @@
+//
+// NOTE: This method opens ports first, and then closes ports. However, it is recommended
+// that your service layer reconciles overlapping port ranges before calling this method.
+func (st *State) UpdateUnitPorts(ctx context.Context, unitUUID string, openPorts, closePorts network.GroupedPortRanges) error {
+	db, err := st.DB()
+	if err != nil {
+		return errors.Trace(err)
+	}
+
+	endpointsUnderAction := set.NewStrings()
+	for endpoint := range openPorts {
+		endpointsUnderAction.Add(endpoint)
+	}
+	for endpoint := range closePorts {
+		endpointsUnderAction.Add(endpoint)
+	}
+
+	getProtocolID, err := st.Prepare("SELECT &protocolID.* FROM protocol WHERE protocol = $protocolName.protocol", protocolID{}, protocolName{})
+	if err != nil {
+		return errors.Annotate(err, "preparing get protocol ID statement")
+	}
+
+	insertPortRange, err := st.Prepare("INSERT INTO port_range (*) VALUES ($unitPortRange.*)", unitPortRange{})
+	if err != nil {
+		return errors.Annotate(err, "preparing insert port range statement")
+	}
+
+	clearPortRange, err := st.Prepare(`
+DELETE FROM port_range
+WHERE unit_endpoint_uuid = $unitEndpointUUID.uuid
+`, unitEndpointUUID{})
+	if err != nil {
+		return errors.Annotate(err, "preparing clear port range statement")
+	}
+
+	return db.Txn(ctx, func(ctx context.Context, tx *sqlair.TX) error {
+
+		unitEndpointMap := map[string]unitEndpointUUID{}
+		for _, endpoint := range endpointsUnderAction.Values() {
+			var err error
+			unitEndpointMap[endpoint], err = st.getUnitEndpointUUID(ctx, tx, unitUUID, endpoint)
+			if errors.Is(err, UnitEndpointNotFound) {
+				unitEndpointMap[endpoint], err = st.insertUnitEndpoint(ctx, tx, unitUUID, endpoint)
+			}
+			if err != nil {
+				return errors.Trace(err)
+			}
+		}
+
+		for endpoint, unitEndpointUUID := range unitEndpointMap {
+			currentOpenedPorts, err := st.getOpenedPorts(ctx, tx, unitEndpointUUID)
+			if err != nil {
+				return errors.Trace(err)
+			}
+
+			currentNetworkPorts := transform.Slice(currentOpenedPorts, decodePortRange)
+			network.SortPortRanges(currentNetworkPorts)
+
+			resultantOpenPorts := reconcilePorts(currentNetworkPorts, openPorts[endpoint], closePorts[endpoint])
+			network.SortPortRanges(resultantOpenPorts)
+
+			// Return early if no changes are to be made. We previously sorted so
+			// port range permutations aren't picked up as changes
+			if reflect.DeepEqual(currentNetworkPorts, resultantOpenPorts) {
+				continue
+			}
+
+			// Clear out all existing open ports for this unit's endpoint and insert the freshly calculated ones.
+			err = tx.Query(ctx, clearPortRange, unitEndpointUUID).Run()
+			if err != nil {
+				return domain.CoerceError(err)
+			}
+
+			for _, portRange := range resultantOpenPorts {
+				var protocolID protocolID
+				err = tx.Query(ctx, getProtocolID, protocolName{Name: portRange.Protocol}).Get(&protocolID)
+				if err != nil {
+					return errors.Annotatef(ProtocolNotValid, "protocol %q", portRange.Protocol)
+				}
```

Comment:
This can be done before the loop, or can be done here if the values haven't been already set.

Reply:
I have fallen back to pulling the protocols out of the map first before the loop and constructing a reverse map from protocol to id, which I then use to later look up ids.

This is the solution which minimises queries, but it does theoretically pull more data out on DQLite than is necessary, since we pull out the id of, for example, icmp when we only open/close tcp.

I think in this scenario the extra readability is worth it, given how little extra data we are reasonably pulling out of state.

Reply:
I have made some major refactors since this original comment. Needs re-evaluating from scratch probably!